dhnx/simulation.py

```diff
@@ -10,23 +10,440 @@
+                    graph_with_data, weight='pipes-mass_flow'
+                ).todense()
+
+                exponent = np.multiply(exponent_constant, inverse_mass_flows)
+
+                matrix = np.exp(exponent)
+
+                # Clear out elements where matrix was zero before exponentiation. This could be
+                # replaced by properly passing `where` to np.multiply in the line above.
+                matrix = np.multiply(matrix, nx.adjacency_matrix(self.nx_graph).todense())
+
+                # Adapt matrix
+                if direction == 1:
+                    matrix = matrix.T
+                    normalisation = np.array(nx.adjacency_matrix(self.nx_graph).sum(0)).flatten()
+
+                    normalisation = np.divide(np.array([1]), normalisation, where=normalisation!=0)
+
+                elif direction == -1:
+                    normalisation = np.array(nx.adjacency_matrix(self.nx_graph).sum(1)).flatten()
+
+                    normalisation = np.divide(np.array([1]), normalisation, where=normalisation!=0)
+
+                else:
+                    raise ValueError("Direction has to be either 1 or -1.")
+
+                normalisation = np.diag(normalisation)
+
+                matrix = np.dot(normalisation, matrix)
+
+                matrix = np.identity(matrix.shape[0]) - matrix
+
+                vector = np.array(known_temp.loc[t])
+
+                vector[vector!=0] -= self.temp_env
+
+                x, residuals, rank, s = np.linalg.lstsq(
+                    matrix,
+                    vector,
+                    rcond=None
+                )
+
+                temps.update({t: x + self.temp_env})
+
+            temp_df = pd.DataFrame.from_dict(
+                temps,
+                orient='index',
+                columns=self.nx_graph.nodes()
+            )
+
+            return temp_df
+
+        temp_inlet = _calc_temps(exponent_constant, self.input_data.temp_inlet, direction=1)
+
+        def _set_temp_return_input(temp_inlet):
+
+            temp_return = pd.DataFrame(
+                0,
+                columns=self.nx_graph.nodes(),
+                index=self.thermal_network.timeindex
+            )
+
+            temp_drop = self._concat_sequences('temperature_drop')
+
+            temp_return.loc[:, temp_drop.columns] = temp_inlet.loc[:, temp_drop.columns] - temp_drop
+
+            return temp_return
+
+        temp_return_known = _set_temp_return_input(temp_inlet)
+
+        temp_return = _calc_temps(exponent_constant, temp_return_known, direction=-1)
+
+        def _calculate_pipes_heat_losses(temp_node):
+
+            pipes_heat_losses = {}
+
+            for i, row in temp_inlet.iterrows():
+
+                mass_flow = self.results['pipes-mass_flow'].loc[i, :].copy()
+
+                temp_difference = np.array(np.dot(row, self.inc_mat)).flatten()
+
+                heat_losses = mass_flow
+
+                pipes_heat_losses[i] = heat_losses.multiply(temp_difference, axis=0)
```

Comment:
Here the mass flow needs to be multiplied with temp_difference and _self.c_

Reply:
Done in 2197872.